tests/src/Module/Api/ApiResponseTest.php

```diff
@@ -0,0 +1,108 @@
+	}
+
+	public function testErrorWithXml()
+	{
+		$l10n = \Mockery::mock(L10n::class);
+		$args = \Mockery::mock(Arguments::class);
+		$args->shouldReceive('getQueryString')->andReturn('');
+
+		$response = new ApiResponseDouble($l10n, $args, new NullLogger());
+		$response->error(200, 'OK', 'error_message', 'xml');
+
+		self::assertEquals('<?xml version="1.0"?>' . "\n" .
+						   '<status xmlns="http://api.twitter.com" xmlns:statusnet="http://status.net/schema/api/1/" ' .
+						   'xmlns:friendica="http://friendi.ca/schema/api/1/" ' .
+						   'xmlns:georss="http://www.georss.org/georss">' . "\n" .
+						   '  <error>error_message</error>' . "\n" .
+						   '  <code>200 OK</code>' . "\n" .
+						   '  <request/>' . "\n" .
+						   '</status>' . "\n",
+			ApiResponseDouble::getOutput());
+	}
+
+	public function testErrorWithRss()
+	{
+		$l10n = \Mockery::mock(L10n::class);
+		$args = \Mockery::mock(Arguments::class);
+		$args->shouldReceive('getQueryString')->andReturn('');
+
+		$response = new ApiResponseDouble($l10n, $args, new NullLogger());
+		$response->error(200, 'OK', 'error_message', 'rss');
+
+		self::assertEquals(
+			'<?xml version="1.0"?>' . "\n" .
+			'<status xmlns="http://api.twitter.com" xmlns:statusnet="http://status.net/schema/api/1/" ' .
+			'xmlns:friendica="http://friendi.ca/schema/api/1/" ' .
+			'xmlns:georss="http://www.georss.org/georss">' . "\n" .
+			'  <error>error_message</error>' . "\n" .
+			'  <code>200 OK</code>' . "\n" .
+			'  <request/>' . "\n" .
+			'</status>' . "\n",
+			ApiResponseDouble::getOutput());
+	}
+
+	public function testErrorWithAtom()
+	{
+		$l10n = \Mockery::mock(L10n::class);
+		$args = \Mockery::mock(Arguments::class);
+		$args->shouldReceive('getQueryString')->andReturn('');
+
+		$response = new ApiResponseDouble($l10n, $args, new NullLogger());
+		$response->error(200, 'OK', 'error_message', 'atom');
+
+		self::assertEquals(
+			'<?xml version="1.0"?>' . "\n" .
+			'<status xmlns="http://api.twitter.com" xmlns:statusnet="http://status.net/schema/api/1/" ' .
+			'xmlns:friendica="http://friendi.ca/schema/api/1/" ' .
+			'xmlns:georss="http://www.georss.org/georss">' . "\n" .
+			'  <error>error_message</error>' . "\n" .
+			'  <code>200 OK</code>' . "\n" .
+			'  <request/>' . "\n" .
+			'</status>' . "\n",
+			ApiResponseDouble::getOutput());
+	}
+
+	public function testUnsupported()
+	{
+		$l10n = \Mockery::mock(L10n::class);
+		$l10n->shouldReceive('t')->andReturnUsing(function ($args) {
+			return $args;
+		});
+		$args = \Mockery::mock(Arguments::class);
+		$args->shouldReceive('getQueryString')->andReturn('');
+
+		$response = new ApiResponseDouble($l10n, $args, new NullLogger());
+		$response->unsupported();
+
+		self::assertEquals('{"error":"API endpoint %s %s is not implemented","error_description":"The API endpoint is currently not implemented but might be in the future."}', ApiResponseDouble::getOutput());
```

Comment:
Can't we do `$response->getOutput`?

Reply:
Because the test classes are currently not always aware of the `ApiResponseDouble` instances.

In this particular case, you're right it would, but in cases where the API calls the `ApiResponse` inside it's method ( with `DI::apiResponse()`), I cannot get the `ApiResponseDouble` per `DI::apiResponse()`, so I would have to ugly force cast `ApiResponse` to `ApiResponseDouble` after calling `DI::apiResponse()` and hope not getting a class cast exception :).

I'm currently thinking about refactoring the `Module` class, so the response (even with `rawOutput` is always a stringyfied content and so I could use the `(Api)Response` logic afterwards. This would make it possible to easily test Modules and Responses without creating doubles or spies for it :-)